chainer/functions/noise/dropout.py

```diff
@@ -34,29 +34,23 @@ def forward_cpu(self, x):
                 and self.mask is None):
             return self._forward_ideep(x)
 
-        if self.mask is not None:
-            y = x[0] * self.mask
-        else:
+        if self.mask is None:
             scale = x[0].dtype.type(1. / (1 - self.dropout_ratio))
             flag = numpy.random.rand(*x[0].shape) >= self.dropout_ratio
             self.mask = scale * flag
-            y = x[0] * self.mask
+        y = x[0] * self.mask
         return y,
 
     def forward_gpu(self, x):
         if (chainer.should_use_cudnn('>=auto', 5000)
-                and x[0].flags.c_contiguous
-                and self.mask is None
-                and not self.return_mask):
+                and x[0].flags.c_contiguous):
             self._use_cudnn = True
 
-            if hasattr(self, 'states'):
-                # if we already have a dropout mask,
-                # the forward operation is equal to backward.
-                return cuda.get_cudnn_dropout_states().backward(
-                    None, x[0], self.dropout_ratio, self.states),
+            # Mask is actually a random seed to ensure
+            # reproducibility within different function calls
+            self.mask, states = cuda.get_cudnn_dropout_states(self.mask)
 
-            self.states, y = cuda.get_cudnn_dropout_states().forward(
+            self.states, y = states.forward(
```

Comment:
It's confusing because `states` and `self.states` are actually very different things.
How about using different naming, e.g. `_cudnn_reserve_space`, following cuDNN/CuPy's naming of the variable?

Reply:
I agree with this.